File: drishti-mvp-functions/getSignedUploadUrl.js
// functions/getSignedUploadUrl.js

const functions = require("firebase-functions");
const { Storage } = require("@google-cloud/storage");
const { v4: uuidv4 } = require("uuid");

// Get centralized dependencies from index.js
const { db } = require('./index.js').dependencies;

const storage = new Storage();
const defaultBucket = "project-drishti-central1-bucket";
const venueBucket = "project-drishti-central1-bucket-venues";

// CORS handler utility
const corsHandler = (req, res, callback) => {
  res.set('Access-Control-Allow-Origin', '*');
  res.set('Access-Control-Allow-Methods', 'GET, POST, PUT, DELETE, OPTIONS');
  res.set('Access-Control-Allow-Headers', 'Content-Type, Authorization');

  if (req.method === 'OPTIONS') {
    res.status(204).send('');
    return;
  }

  callback();
};

module.exports = functions.https.onRequest(async (req, res) => {
  corsHandler(req, res, async () => {
    try {
      const { filename, mimetype, zone, notes, type, bucket } = req.body;

      if (!filename || !mimetype || !zone) {
        return res.status(400).json({ error: "Missing required fields." });
      }

      const fileId = uuidv4();
      const objectPath = `uploads/${fileId}_${filename}`;

      // Choose bucket based on request
      let bucketName;
      if (bucket) {
        bucketName = bucket;
      } else if (type === "venue") {
        bucketName = venueBucket;
      } else {
        bucketName = defaultBucket;
      }

      const options = {
        version: "v4",
        action: "write",
        expires: Date.now() + 5 * 60 * 1000, // 5 minutes
        contentType: mimetype,
      };

      const [url] = await storage
        .bucket(bucketName)
        .file(objectPath)
        .getSignedUrl(options);

      // Store metadata
      await db.collection("media").add({
        zone,
        notes: notes || "",
        fileUrl: `https://storage.googleapis.com/${bucketName}/${objectPath}`,
        objectPath,
        type: mimetype.startsWith("image") ? "image" : "video",
        timestamp: new Date(),
      });

      res.status(200).json({ url, objectPath });
    } catch (err) {
      console.error("Error generating signed URL:", err);
      res.status(500).json({ error: `Failed to generate signed URL: ${err.message}` });
    }
  });
});

File: drishti-mvp-adk-service/get_signed_upload_url.py
import os
from fastapi import APIRouter, HTTPException
from typing import Dict
from google.cloud import storage, firestore
from datetime import timedelta, datetime
import uuid

router = APIRouter()

default_bucket = os.getenv("GCS_BUCKET_NAME", "project-drishti-central1-bucket")
venue_bucket = os.getenv("GCS_VENUE_BUCKET_NAME", "project-drishti-central1-bucket-venues")
vision_bucket = os.getenv("GCS_VISION_BUCKET_NAME", "project-drishti-central1-bucket-vision")

db = firestore.Client()

@router.post("/media/upload-url")
def get_signed_upload_url(request: Dict):
    filename = request.get("filename")
    content_type = request.get("content_type", "application/octet-stream")
    zone = request.get("zone")
    notes = request.get("notes", "")
    file_type = request.get("type")
    bucket_override = request.get("bucket")

    if not filename or not zone:
        raise HTTPException(status_code=400, detail="Filename and zone are required.")

    # Determine bucket
    if bucket_override:
        bucket_name = bucket_override
    elif file_type == "venue":
        bucket_name = venue_bucket
    elif file_type == "vision":
        bucket_name = vision_bucket
    else:
        bucket_name = default_bucket

    file_id = str(uuid.uuid4())
    object_path = f"uploads/{file_id}_{filename}"

    try:
        storage_client = storage.Client()
        bucket = storage_client.bucket(bucket_name)
        blob = bucket.blob(object_path)
        url = blob.generate_signed_url(
            version="v4",
            expiration=timedelta(minutes=15),
            method="PUT",
            content_type=content_type,
        )
        # Store metadata in Firestore
        file_url = f"https://storage.googleapis.com/{bucket_name}/{object_path}"
        db.collection("media").add({
            "zone": zone,
            "notes": notes,
            "fileUrl": file_url,
            "objectPath": object_path,
            "type": "image" if content_type.startswith("image") else "video",
            "timestamp": datetime.utcnow(),
        })
        return {"uploadUrl": url, "objectPath": object_path}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e)) 